src/meshes/serial_tree.jl

```diff
@@ -54,43 +54,51 @@ mutable struct SerialTree{NDIMS} <: AbstractTree{NDIMS}
         t.child_ids = fill(typemin(Int), 2^NDIMS, capacity + 1)
         t.neighbor_ids = fill(typemin(Int), 2 * NDIMS, capacity + 1)
         t.levels = fill(typemin(Int), capacity + 1)
-        t.coordinates = fill(NaN, NDIMS, capacity + 1)
+        t.coordinates = fill(convert(RealT, NaN), NDIMS, capacity + 1) # `NaN` is of type Float64
         t.original_cell_ids = fill(typemin(Int), capacity + 1)
 
         t.capacity = capacity
         t.length = 0
         t.dummy = capacity + 1
 
-        t.center_level_0 = SVector(ntuple(_ -> NaN, NDIMS))
-        t.length_level_0 = NaN
+        t.center_level_0 = SVector(ntuple(_ -> convert(RealT, NaN), NDIMS))
+        t.length_level_0 = convert(RealT, NaN)
 
         return t
     end
 end
 
-# Constructor for passing the dimension as an argument
-SerialTree(::Val{NDIMS}, args...) where {NDIMS} = SerialTree{NDIMS}(args...)
+# Constructor for passing the dimension as an argument. Default datatype: Float64
+SerialTree(::Val{NDIMS}, args...) where {NDIMS} = SerialTree{NDIMS, Float64}(args...)
 
 # Create and initialize tree
-function SerialTree{NDIMS}(capacity::Int, center::AbstractArray{Float64},
-                           length::Real, periodicity = true) where {NDIMS}
+function SerialTree{NDIMS, RealT}(capacity::Int, center::AbstractArray{RealT},
+                                  length::RealT,
+                                  periodicity = true) where {NDIMS, RealT <: Real}
     # Create instance
-    t = SerialTree{NDIMS}(capacity)
+    t = SerialTree{NDIMS, RealT}(capacity)
 
     # Initialize root cell
     init!(t, center, length, periodicity)
 
     return t
 end
 
+function SerialTree{NDIMS}(capacity::Int, center::AbstractArray{RealT},
+                           length::RealT,
+                           periodicity = true) where {NDIMS, RealT <: Real}
+    t = SerialTree{NDIMS, RealT}(capacity, center, length, periodicity)
+end
+
 # Constructor accepting a single number as center (as opposed to an array) for 1D
-function SerialTree{1}(cap::Int, center::Real, len::Real, periodicity = true)
-    SerialTree{1}(cap, [convert(Float64, center)], len, periodicity)
+function SerialTree{1, RealT}(cap::Int, center::RealT, len::RealT,
+                              periodicity = true) where {RealT <: Real}
+    SerialTree{1, RealT}(cap, [center], len, periodicity)
```

Comment:
We should also have an interface `SerialTree{1}` just like what you have done for `SerialTree{NDIMS}` otherwise I doubt there would be some errors in the precompile here https://github.com/trixi-framework/Trixi.jl/blob/179a2e4a680803909974eeb4c88dabfbb04cc1e3/src/auxiliary/precompile.jl#L397-L405

Reply:
Ah good catch! I will add the corresponding constructor for consistency.